source/mesh_deformation/fastscapecc.cc

```diff
@@ -0,0 +1,550 @@
+                    {
+                      for (unsigned int vertex = 0; vertex < GeometryInfo<dim - 1>::vertices_per_face; ++vertex)
+                        {
+                          const unsigned int dof_index = cell->face(face)->vertex_dof_index(vertex, 0);
+                          boundary_solution[dof_index] = this->get_solution()[dof_index];
+                        }
+                    }
+                }
+            }
+        }
+
+      boundary_solution.compress(VectorOperation::insert);
+
+      // dealii::Functions::FEFieldFunction<2, dealii::LinearAlgebra::distributed::Vector<double>,3>
+      //       fe_function(this->get_dof_handler(), boundary_solution, this->get_mapping());
+
+      // VectorTools::interpolate_boundary_values (surface_mesh_dof_handler,
+      //                                           *boundary_ids.begin(),
+      //                                           boundary_solution,
+      //                                           surface_constraints);
+
+      // VectorTools::interpolate_to_different_mesh(
+      //     this->get_dof_handler(),         // Source DoFHandler
+      //     this->get_solution(),            // Source solution vector
+      //     surface_mesh_dof_handler,        // Target DoFHandler
+      //     surface_solution                 // Target solution vector
+      // );
+
+      // Project the boundary solution onto the 2D surface mesh
+      // VectorTools::project(
+      //     this->template get_mapping<2,3>(),
+      //     surface_mesh_dof_handler,
+      //     surface_constraints,
+      //     quadrature,
+      //     fe_function,
+      //     surface_solution);
+    }
+
+
+
+    template <int dim>
+    void FastScapecc<dim>::compute_velocity_constraints_on_boundary(const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                                    AffineConstraints<double> &mesh_velocity_constraints,
+                                                                    const std::set<types::boundary_id> &boundary_ids) const
+    {
+      if (this->get_timestep_number() == 0)
+        return;
+
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "FastScape plugin");
+
+      const_cast<FastScapecc<dim> *>(this)->project_surface_solution(boundary_ids);
+
+      // VectorTools::interpolate_to_different_mesh(
+      //     this->get_dof_handler(),         // Source DoFHandler
+      //     this->get_solution(),            // Source solution vector
+      //     surface_mesh_dof_handler,        // Target DoFHandler
+      //     surface_solution                 // Target solution vector
+      // );
+
+      // Apply hanging node constraints to ensure continuity
+      // TODO: this yields a link error
+      // surface_constraints.distribute(surface_solution);
+
+      // Temporary storage for computation
+      std::vector<std::vector<double>> temporary_variables(dim + 2, std::vector<double>());
```

Comment:
nit: using nested vectors seems unnecessary in my opinion. We could use separate variables instead, which is more readable.